### PR TITLE
Fix bug on slope and vector ruggedness vector

### PR DESCRIPTION
**Pull request by rthedin**
Fix for the Vector Ruggedness Vector function. Few items changed:
- `slope` calculation now requests `attrib='slope_degrees'`.
- All computations of slope and VRM require knowledge of the resolution in which the height array is given. That should have been given using the `slope_zscale` parameter, but this was far from being clear. Now an explicit (and required) resolution parameter `res` need to be specified.
- Add bool option for filling depressions (default `True`)
- Added option to return the aspect as well

Tested with simple pyramid geometry where the slope can be quickly inferred. Compared using the `mmctools.coupling.terrain.calc_slope` function.
image

A more realistic terrain:
BEFORE:
image
AFTER:
image

### CODE REVIEW

File: mmctools/helper_functions.py [hunk 1]
```diff
@@ -1353,18 +1356,22 @@ def calcVRM(hgt,window=None,footprint=None,slope_zscale=1.0,return_slope=False):
...
+    # Determine scale based on resolution of hgt array
+    zscale = 1/res
+
     # Get slope and aspect:
     hgt_rd = rd.rdarray(hgt, no_data=-9999)
-    rd.FillDepressions(hgt_rd, in_place=True)
-    slope  = rd.TerrainAttribute(hgt_rd, attrib='slope_riserun', zscale=slope_zscale)
+    if fill_depressions:
+        rd.FillDepressions(hgt_rd, in_place=True)
+    slope  = rd.TerrainAttribute(hgt_rd, attrib='slope_degrees',zscale=zscale)
```
Comment (ewquon) [inline mmctools/helper_functions.py:1366]: For backwards compatibility, should we make this a kwarg, with the default being `slope_riserun`? Or add a "degrees" flag and set the attrib accordingly?

**Comment by ewquon** [comment]
And to be clear, `coupling.terrain.calc_slope` works properly, but just excludes boundary pts?

**Comment by ewquon** [comment]
@rthedin, @phawbeck thoughts on moving `calc_VRI` and `calcTRI` to `coupling.terrain`, or alternatively, moving calc_slope to `helper_functions`? My preference is for the former, as helper_functions is a bit bloated

File: mmctools/helper_functions.py (hunk 1 above)
Comment (rthedin) [reply mmctools/helper_functions.py:1366]: I'm not opposed to a flag (even though I see no use case), but `riserun` here would be incorrect. The computation of `rugz` and `rugdxy` depends on the slope _angle_, not rise-run ($\tan^{-1}$)

**Comment by rthedin** [comment]
> And to be clear, `coupling.terrain.calc_slope` works properly, but just excludes boundary pts?

That's right. Note that `calc_slope` is hard-coded for a 3x3 stencil with second-order weights, while `richdem` allows a generic specification of stencil and/or weights through the `footprint` variable. 


> @rthedin, @phawbeck thoughts on moving `calc_VRI` and `calcTRI` to `coupling.terrain`, or alternatively, moving calc_slope to `helper_functions`? My preference is for the former, as helper_functions is a bit bloated

I think that is a good idea. My vote is also for the former.

File: mmctools/helper_functions.py (hunk 1 above)
Comment (ewquon) [reply mmctools/helper_functions.py:1366]: Oh I see. Let's just leave it as is, then.

**Comment by ewquon** [comment]
Oh yeah, I see I put that info right in the comments https://github.com/a2e-mmc/mmctools/blob/dev/mmctools/coupling/terrain.py#L479

**Review by phawbeck — COMMENTED** [review]
I think I see what you're doing here, but I'm not seeing the bug - isn't this just a different way of doing the same thing + you get the borders (which include assumptions)? As long as the units in your x, y, and z variables are 1-to-1, then the original way should calculate slope the same as specifying the resolution. I'm not against the modification, to be clear, but I don't see what it changes.

In terms of moving them, @ewquon, move as you please!

I'm glad this is getting attention from somebody else, though - glad to see the time wasn't wasted on one little side-effort :-)

**Comment by ewquon** [comment]
> I think I see what you're doing here, but I'm not seeing the bug - isn't this just a different way of doing the same thing + you get the borders (which include assumptions)?

I think the issue is the use of rise/run in angle calculations, instead of the slope angle.

**Comment by rthedin** [comment]
Two issues: as Eliot mentioned, the angles going into the calculation of the VRM is rise/run instead of slope; and without knowledge of the resolution of the `hgt` array given, there's no way to compute the "run", only the "rise".

**Comment by phawbeck** [comment]
> I think the issue is the use of rise/run in angle calculations, instead of the slope angle.

Ahhh I see! That sucks for my old work... Yeah, this looks good!

**Comment by phawbeck** [comment]
Should we merge this and then move, or move then merge? Who wants to do the moving? @ewquon @rthedin

**Comment by rthedin** [comment]
I'll push another commit to this PR moving the functions around, and then you guys merge.

**Comment by phawbeck** [comment]
Before we merge this - I pulled the changes into my repo and tested it with my terrain data and am getting slope values of over 200˚ and am not sure how... can you share the code that you used to test this and I'll swap out your terrain for mine?

**Comment by rthedin** [comment]
This is a little snipped for the pyramid that I showed.

```
def plotSlope (xx, yy, zz, sl, slope):
    fig, axs = plt.subplots(ncols=3, nrows=1, figsize=(21, 4.5))

    cm = axs[0].pcolormesh(xx, yy, zz, cmap='plasma', shading='auto')
    cb = plt.colorbar(cm, ax= axs[0], label='height (m)')
    axs[0].set_title('elevation')

    cm = axs[1].pcolormesh(xx, yy, sl, shading='auto')
    cb = plt.colorbar(cm, ax= axs[1], label='slope (deg)')
    axs[1].set_title('mmctools.coupling.terrain.calc_slope function')

    cm = axs[2].pcolormesh(xx, yy, slope, shading='auto')
    cb = plt.colorbar(cm, ax= axs[2], label='slope (deg)')
    axs[2].set_title('richdem function')

    for ax in axs:
        ax.set_aspect('equal', 'box')
        ax.set_xlabel('x (m)'); ax.set_ylabel('y (m)')

# Create pyramid geometry
hmax = 40
zpyr =np.array( [[0,	0	,	0	,	0	,	0	,	0	,	0	,	0	,	0	,	0	,	0],
                 [0,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0],
                 [0,	0.25,	0.5	,	0.5	,	0.5	,	0.5	,	0.5	,	0.5	,	0.5	,	0.25,	0],
                 [0,	0.25,	0.5	,	0.75,	0.75,	0.75,	0.75,	0.75,	0.5	,	0.25,	0],
                 [0,	0.25,	0.5	,	0.75,	1	,	1	,	1	,	0.75,	0.5	,	0.25,	0],
                 [0,	0.25,	0.5	,	0.75,	0.75,	0.75,	0.75,	0.75,	0.5	,	0.25,	0],
                 [0,	0.25,	0.5	,	0.5	,	0.5	,	0.5	,	0.5	,	0.5	,	0.5	,	0.25,	0],
                 [0,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0.25,	0],
                 [0,	0	,	0	,	0	,	0	,	0	,	0	,	0	,	0	,	0	,	0]])
zpyr = hmax * zpyr
xpyr = np.arange(0,101,10)
ypyr = np.arange(0,81, 10)
xxpyr, yypyr = np.meshgrid(xpyr, ypyr, indexing='ij')


# calculate quantities using both methods
_, slope, _ = calcVRM(zpyr, res=10, footprint=np.ones((3,3)), fill_depressions=False, return_slope_aspect=True)
sl = calc_slope(xxpyr, yypyr, zpyr)

plotSlope(xpyr, ypyr, zpyr, sl, slope)
```

**Comment by rthedin** [comment]
The real terrain case is not simple for you to reproduce. I have another simple yet not-so-simple case of a cos bump. You can see how the slopes are reasonable.

```
def createBump (L, h, res=10):
     
    # Extents of the outer domain. Little larger bc of openfoam
    xMIN = -500.0; yMIN = -500.0; 
    xMAX =  500.0; yMAX =  500.0;
    
    # Mesh parameters
    nx = int(np.ceil((xMAX-xMIN)/res)) +1
    ny = int(np.ceil((yMAX-yMIN)/res)) +1
    
    # Create meshgrid
    x = np.linspace(xMIN, xMAX, nx)
    y = np.linspace(yMIN, yMAX, ny)
    xx, yy = np.meshgrid(x,y, indexing='ij')

    # Create bump
    zbump = h * np.cos(np.pi*(yy + xx)/(2*L))**2
    zbump[abs(yy +xx) > L]=0
    
        
    # calculate quantities using both methods
    _, slope, _ = calcVRM(zbump, res=res, footprint=np.ones((3,3)), fill_depressions=False, return_slope_aspect=True)
    sl = calc_slope(xx, yy, zbump)

    return xx, yy, zbump, slope, sl

xx, yy, zbump, slope, sl = createBump (L=100, h=40)
plotSlope(xx,yy,zbump,sl,slope)
```

Running it, you get:
image

**Comment by phawbeck** [comment]
As long as for your terrain case, you're just swapping out `zbump` with your terrain array, then I can see what I get from it with my terrain. I don't need your terrain, just the command for calculating slope.

**Comment by phawbeck** [comment]
Using your test case with my terrain worked - must still be some old artifacts in my script that adjust slope. I think we're good. @ewquon @rthedin

**Comment by rthedin** [comment]
Awesome. Thanks for checking it with your data as well! And yes, I'm just swapping the `zbump` with my elevation array.

**Pull request merged**